docs/source/developer/contributing.rst

```diff
@@ -34,66 +642,179 @@ Writing Style
+of the packages in the repo. It will:
+
+-  Ensure the core package version dependencies match everywhere.
+-  Ensure imported packages match dependencies.
+-  Ensure a consistent version of all packages.
+-  Manage the meta package.
+
+The ``packages/metapackage`` package is used to build all of the
+TypeScript in the repository at once, instead of 50+ individual builds.
+
+The integrity script also allows you to automatically add a dependency
+for a package by importing from it in the TypeScript file, and then
+running: ``jlpm run integrity`` from the repo root.
+
+We also have scripts for creating and removing packages in
+``packages/``, ``jlpm run create:package`` and
+``jlpm run remove:package``. When creating a package, if it is meant to
+be included in the core bundle, add the
+``jupyterlab: { coreDependency: true }`` metadata to the
+``package.json``. Packages with ``extension`` or ``mimeExtension``
+metadata are considered to be a core dependency unless they are
+explicitly marked otherwise.
+
+Testing Changes to External Packages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Linking/Unlinking Packages to JupyterLab
+''''''''''''''''''''''''''''''''''''''''
+
+If you want to make changes to one of JupyterLab's external packages
+(for example, `Lumino <https://github.com/jupyterlab/lumino>`__ and test
+them out against your copy of JupyterLab, you can easily do so using the
+``link`` command:
+
+1. Make your changes and then build the external package
+2. Register a link to the modified external package
+
+   -  navigate to the external package dir and run ``jlpm link``
+
+3. Link JupyterLab to modded package
+
+   -  navigate to top level of your JupyterLab repo, then run
+      ``jlpm link "<package-of-interest>"``
+
+You can then (re)build JupyterLab (eg ``jlpm run build``) and your
+changes should be picked up by the build.
+
+To restore JupyterLab to its original state, you use the ``unlink``
+command:
+
+1. Unlink JupyterLab and modded package
+
+   -  navigate to top level of your JupyterLab repo, then run
+      ``jlpm unlink "<package-of-interest>"``
+
+2. Reinstall original version of the external package in JupyterLab
+
+   -  run ``jlpm install --check-files``
+
+You can then (re)build JupyterLab and everything should be back to
+default.
+
+Possible Linking Pitfalls
+'''''''''''''''''''''''''
+
+If you're working on an external project with more than one package,
+you'll probably have to link in your copies of every package in the
+project, including those you made no changes to. Failing to do so may
+cause issues relating to duplication of shared state.
+
+Specifically, when working with Lumino, you'll probably have to link
+your copy of the ``"@lumino/messaging"`` package (in addition to
+whatever packages you actually made changes to). This is due to
+potential duplication of objects contained in the ``MessageLoop``
+namespace provided by the ``messaging`` package.
+
 User Interface Naming Conventions
```

Comment:
The "User Interface Naming Conventions" section and all of its subsections should be under the "Writing Document" section

Reply:
probably put it after the "Writing Style" subsection